cpp/src/arrow/builder.cc

```diff
@@ -43,6 +44,33 @@ namespace arrow {
 
 using internal::AdaptiveIntBuilderBase;
 
+namespace {
+
+Status TrimBuffer(const int64_t bytes_filled, ResizableBuffer* buffer) {
+  if (buffer) {
+    if (bytes_filled < buffer->size()) {
+      // Trim buffer
+      RETURN_NOT_OK(buffer->Resize(bytes_filled));
+    }
+    // zero the padding
+    buffer->ZeroPadding();
+  } else {
+    DCHECK_EQ(bytes_filled, 0);
+  }
+  return Status::OK();
+}
+
+}  // namespace
+
+ArrayBuilder::~ArrayBuilder() {
+#ifndef NDEBUG
+  if (ARROW_PREDICT_FALSE(!is_finished_)) {
```

Comment:
Is this desirable? If an error occurs while building an array, it is normal not to finish the builder.

Reply:
How about hiding this behind a flag like `ARROW_LOG_UNFINISHED_BUILDERS`. Having this output in debug builds always seems like too much to me

Reply:
You want to rework it and move the warning to `null_bitmap` and `data` methods. It seems to be doable, but I'd need some more time. 

We can still then move it behind a special compile flag.